app/controllers/spree/orders_controller_decorator.rb

```diff
@@ -130,6 +131,36 @@ def cancel
 
   private
 
+  # Updates the various denormalized total attributes of the order and
+  # recalculates the shipment taxes
+  def update_totals_and_taxes
+    @order.updater.update_totals
+    @order.shipment.ensure_correct_adjustment_with_included_tax if @order.shipment
+  end
+
+  # Sets the adjustments to open to perform the block's action and restores
+  # their state to whatever the they had. Note that it does not change any new
+  # adjustments that might get created in the yielded block.
+  def with_open_adjustments
+    previous_states = @order.adjustments.each_with_object({}) do |adjustment, hash|
+      hash[adjustment.id] = { adjustment: adjustment, previous_state: adjustment.state }
+    end
+    @order.adjustments.each(&:open)
+
+    yield
+
+    previous_states.each do |adjustment_id, adjustment_pair|
+      adjustment = adjustment_pair[:adjustment]
+      previous_state = adjustment_pair[:previous_state]
+
+      adjustment.update_attribute(:state, previous_state)
+    end
+  end
```

Comment:
```suggestion
  def with_open_adjustments
    previous_states = build_adjustments_state_hash
    @order.adjustments.each(&:open)
    yield
    restore_adjustments_state(previous_states)
  end
  
  def build_adjustments_state_hash
     @order.adjustments.each_with_object({}) do |adjustment, hash|
      hash[adjustment.id] = { adjustment: adjustment, previous_state: adjustment.state }
    end
  end
  
  def restore_adjustments_state(adjustments_state_hash)
    adjustments_state_hash.each do |adjustment_id, adjustment_pair|
      adjustment = adjustment_pair[:adjustment]
      previous_state = adjustment_pair[:previous_state]

      adjustment.update_attribute(:state, previous_state)
    end
  end
```

Reply:
I'd extract method ;-)

Reply:
It's hairy I agree. I'd prefer to extract class but in another PR as IMO this is growing too much and chances are it'll take some time. Agree? If so, I'll create an issue.

Reply:
sure!
if this was my PR I'd extract method and leave it there. extract class is even better but will take more time. it's really up to you.

Reply:
that's right, but when I see the rest of the controller I feel really bad adding a new method...

Reply:
Done in https://github.com/openfoodfoundation/openfoodnetwork/issues/3376

Reply:
:+1: